csrc/megatron/fused_rotary_positional_embedding.h

```diff
@@ -367,3 +441,46 @@ void dispatch_fused_rope_thd_backward(
       o_stride_d, output_grads, cu_seqlens, freqs, input_grads);
   C10_CUDA_KERNEL_LAUNCH_CHECK();
 }
+
+template <typename scalar_t_0, typename scalar_t_1>
+void dispatch_fused_rope_2d_forward(
+    const int b, const int ih, const int iw, const int h, const int d,
+    const int stride_b, const int stride_ih, const int stride_iw,
+    const int stride_h, const int stride_d, const int o_stride_b,
+    const int o_stride_s, const int o_stride_h, const int o_stride_d,
+    const scalar_t_0* input, const scalar_t_1* cos_h, const scalar_t_1* sin_h,
+    const scalar_t_1* cos_w, const scalar_t_1* sin_w, scalar_t_0* output) {
+  auto stream = at::cuda::getCurrentCUDAStream();
+
+  int warps_per_block = h < 16 ? 4 : 8;
+  dim3 blocks(ih, iw, b);
+  dim3 threads(C10_WARP_SIZE, warps_per_block);
+
+  fused_rope_2d_forward<<<blocks, threads, 0, stream>>>(
+      ih, iw, h, d, stride_b, stride_ih, stride_iw, stride_h, stride_d,
+      o_stride_b, o_stride_s, o_stride_h, o_stride_d, input, cos_h, sin_h,
+      cos_w, sin_w, output);
+  C10_CUDA_KERNEL_LAUNCH_CHECK();
+}
+
+template <typename scalar_t_0, typename scalar_t_1>
+void dispatch_fused_rope_2d_backward(
+    const int b, const int ih, const int iw, const int h, const int d,
+    const int stride_b, const int stride_ih, const int stride_iw,
+    const int stride_h, const int stride_d, const int o_stride_b,
+    const int o_stride_s, const int o_stride_h, const int o_stride_d,
+    const scalar_t_0* output_grads, const scalar_t_1* cos_h,
+    const scalar_t_1* sin_h, const scalar_t_1* cos_w, const scalar_t_1* sin_w,
+    scalar_t_0* input_grads) {
+  auto stream = at::cuda::getCurrentCUDAStream();
+
+  int warps_per_block = h < 16 ? 4 : 8;
+  dim3 blocks(ih, iw, b);
+  dim3 threads(C10_WARP_SIZE, warps_per_block);
+
+  fused_rope_2d_backward<<<blocks, threads, 0, stream>>>(
+      ih, iw, h, d, stride_b, stride_ih, stride_iw, stride_h, stride_d,
+      o_stride_b, o_stride_s, o_stride_h, o_stride_d, output_grads, cos_h,
+      sin_h, cos_w, sin_w, input_grads);
+  C10_CUDA_KERNEL_LAUNCH_CHECK();
+}
```

Comment:
could you have this file end with a new line?

Reply:
Resolved.